src/header-generator.js

```diff
@@ -0,0 +1,310 @@
+    httpVersion: ow.optional.string,
+};
+
+const headerGeneratorOptionsShape = {
+    browsers: ow.optional.array.ofType(ow.object.exactShape(browserSpecificationShape)),
+    operatingSystems: ow.optional.array.ofType(ow.string),
+    devices: ow.optional.array.ofType(ow.string),
+    locales: ow.optional.array.ofType(ow.string),
+    httpVersion: ow.optional.string,
+};
+
+/**
+ * @typedef BrowserSpecification
+ * @param {string} name - One of `chrome`, `firefox` and `safari`.
+ * @param {number} minVersion - Minimal version of browser used.
+ * @param {number} maxVersion - Maximal version of browser used.
+ * @param {string} httpVersion - Http version to be used to generate headers (the headers differ depending on the version).
+ *  Either 1 or 2. If none specified the httpVersion specified in `HeaderGeneratorOptions` is used.
+ */
+/**
+ * @typedef HeaderGeneratorOptions
+ * @param {Array<BrowserSpecification>} browsers - List of BrowserSpecifications to generate the headers for.
+ * @param {Array<string>} operatingSystems - List of operating systems to generate the headers for.
+ *  The options are `windows`, `macos`, `linux`, `android` and `ios`.
+ * @param {Array<string>} devices - List of devices to generate the headers for. Options are `desktop` and `mobile`.
+ * @param {Array<string>} locales - List of at most 10 languages to include in the
+ *  [Accept-Language](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language) request header
+ *  in the language format accepted by that header, for example `en`, `en-US` or `de`.
+ * @param {string} httpVersion - Http version to be used to generate headers (the headers differ depending on the version).
+ *  Can be either 1 or 2. Default value is 2.
+ */
+
+/**
+ * HeaderGenerator randomly generates realistic browser headers based on specified options.
+ */
+class HeaderGenerator {
+    /**
+     * @param {HeaderGeneratorOptions} options - default header generation options used unless overridden
+     */
+    constructor(options = {}) {
+        ow(options, 'HeaderGeneratorOptions', ow.object.exactShape(headerGeneratorOptionsShape));
+        this.defaultOptions = options;
+        this.headersOrder = JSON.parse(fs.readFileSync(headersOrderFilePath, { encoding: 'utf8' }));
+        const uniqueBrowserStrings = JSON.parse(fs.readFileSync(browserHelperFilePath, { encoding: 'utf8' }));
+        this.uniqueBrowsers = [];
+        for (const browserString of uniqueBrowserStrings) {
+            if (browserString === missingValueDatasetToken) {
+                this.uniqueBrowsers.push({
+                    name: missingValueDatasetToken,
+                });
+            } else {
+                this.uniqueBrowsers.push(prepareHttpBrowserObject(browserString));
+            }
+        }
+        this.inputGeneratorNetwork = new BayesianNetwork(inputNetworkDefinitionPath);
+        this.headerGeneratorNetwork = new BayesianNetwork(headerNetworkDefinitionPath);
```

Comment:
Same here. Does not make sense to load and parse the files repeatedly. We should use the loaded definition instead of the path, or we need to cache the path / object in the `BayesianNetwork` module. I would just use the loaded object as the parameter.

It's often cumbersome to work with constructors that accept paths, because if you want to programmatically change something in the input, you need to write the changes to a file instead of just modifying an object and be done with it.

Reply:
What you are suggesting goes against the fundamental principle of encapsulation. The HeaderGenerator class has no business knowing anything about the definitions of the BayesianNetwork, or even worse, changing them.
However, as mentioned before, we can certainly preload them if you feel strongly about that part.

Reply:
I don't think there's a difference in encapsulation in those two formats:

```js
const foo = new Dog(pathToConfig);
const bar = new Cat(config);
```

Neither needs to know anything about the config. It just needs to provide it in the correct format. Both are totally fine ways of doing things. But we need to make sure they perform well.

We can either:

```js
const networkDefinition = require('./network-definition.json');
// ...
const foo = new BayesianNetwork(networkDefinition);
```

or we need to implement lazy loading in the `BayesianNetwork`. Both are good solutions. The lazy loading is probably a bit better, but also more time consuming.

Reply:
I meant the "It's often cumbersome to work with constructors that accept paths, because if you want to programmatically change something in the input, you need to write the changes to a file instead of just modifying an object and be done with it." comment, not the options of loading. As I added afterwards, those are ok.

Reply:
Yeah, I should've just left that second part of the comment out. It was only confusing.